hack/.ci/lib/e2e.sh

```diff
@@ -0,0 +1,253 @@
+#!/bin/bash
+#
+# Copyright (C) 2024 ScyllaDB
+#
+
+set -euExo pipefail
+shopt -s inherit_errexit
+
+source "$( dirname "${BASH_SOURCE[0]}" )/../../lib/kube.sh"
+
+# gather-artifacts is a self sufficient function that collects artifacts without depending on any external objects.
+function gather-artifacts {
+  kubectl create namespace gather-artifacts --dry-run=client -o=yaml | kubectl_apply -f=-
+  kubectl create clusterrolebinding gather-artifacts --clusterrole=cluster-admin --serviceaccount=gather-artifacts:default --dry-run=client -o=yaml | kubectl_apply -f=-
+  kubectl create -n=gather-artifacts pdb must-gather --selector='app=must-gather' --max-unavailable=0 --dry-run=client -o=yaml | kubectl_apply -f=-
+
+  kubectl_create -n=gather-artifacts -f=- <<EOF
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    app: must-gather
+  name: must-gather
+spec:
+  restartPolicy: Never
+  containers:
+  - name: wait-for-artifacts
+    command:
+    - /usr/bin/sleep
+    - infinity
+    image: "${SO_IMAGE}"
+    imagePullPolicy: Always
+    volumeMounts:
+    - name: artifacts
+      mountPath: /tmp/artifacts
+  - name: must-gather
+    args:
+    - must-gather
+    - --all-resources
+    - --loglevel=2
+    - --dest-dir=/tmp/artifacts
+    image: "${SO_IMAGE}"
+    imagePullPolicy: Always
+    volumeMounts:
+    - name: artifacts
+      mountPath: /tmp/artifacts
+  volumes:
+  - name: artifacts
+    emptyDir: {}
+EOF
+  kubectl -n=gather-artifacts wait --for=condition=Ready pod/must-gather
+
+  exit_code="$( wait-for-container-exit-with-logs gather-artifacts must-gather must-gather )"
+
+  kubectl -n=gather-artifacts cp --retries=42 -c=wait-for-artifacts must-gather:/tmp/artifacts "${ARTIFACTS}/must-gather"
```

Comment:
validate if ARTIFACTS is non empty at the beginning of the function

Reply:
This can technically be empty for top level dir, but it shall not be undefined which is already enforced

Reply:
ARTIFACTS isn't validated but SO_IMAGE which isn't used in this function

Reply:
SO_IMAGE is used in this function (gather-artifacts) on line 47
https://github.com/scylladb/scylla-operator/pull/1901/files#diff-3d1050b6af0547d51c2384888158e1cdbd2645db709e7af5f737d7f1bcefcd41R47

ARTIFACTS not set is validated by setting shell exit on undefined variable (that catches furgetting to set the variable), ARTIFACTS="" results in "/must-gather" which is a valid path